### PR TITLE
Remove global singletons from codebase

### PR DESCRIPTION
**Pull request by awygle**
This pull request refactors the Config and CoreManager classes so they are no longer singletons, and are instead passed to the functions and objects which need them. This is intended to make testing and future development easier.

I might have gone a little overboard here - if you wanted CoreManager to remain a singleton, then only the first two commits are needed.

### CODE REVIEW

**Comment by olofk** [comment]
FuseSoC is all about excess and living life to it's fullest, so going overboard is not an issue. Thanks for the patches. I've glanced through some of them quickly and they seem fine, but I'll try to give a proper review this weekend

**Comment by olofk** [comment]
Some comments. Overall it looks great and makes the codebase tidier, but there are some things I would like to see changed:

1. This is the big one. I'm afraid the first patch will break the cygpath setting in the opencores provider. Not sure how to deal with that to be honest. It's a bit awkward to send the config all the way down to the providers for just this one setting in one provider. Would like to have some input from @cr1901 if this should be done for all providers when `--cygpath` is enabled or if it's specific to svn. I'm inclined to move the cygpath handling into core.py

2. Make cd2c243e82039d2c2e66cbc78b581a97dda11893 a part of 9d7de917d86c2fe1bab5b84f7c554d8e4c7a4db0

3. Make 747e74f2d9e1d50cc34238052443eb67360e903c a part of 238c287cabe79e9f94b85b5b31de9610e8724460

**Comment by cr1901** [comment]
>Would like to have some input from @cr1901 if this should be done for all providers when --cygpath is enabled or if it's specific to svn.

TLDR: Revert my changes until I think of a better solution.

Long story: The more I think about it, the more I'm inclined to feel my solution is a hack.

My philosophy on dealing with the mess on Windows is something like this (substitute `X` for `cygwin`, `msys2`, `native`):
* If using `X`, use the `X` `python` and call out to `X` binaries as much as possible. If calling out to an `X` binary is not possible, try to constrain abstraction leaks as much as possible. See [this issue](https://github.com/m-labs/migen/pull/88) for an example.

One big abstraction leak when mixing binaries from various environments in Windows is path handling. In my case, the `msys2` `svn` doesn't really exist; the `svn` that `msys2` provides is a `cygwin`-like binary that only understands POSIX paths. An `msys2` `svn` would understand both POSIX and Windows paths. So to contain the abstraction leak, I think we should only add [platform-specific logic](https://stackoverflow.com/questions/35956180/how-to-determine-the-python-is-mingw-or-official-build) to the OpenCores builder only to handle path conversions.

**Comment by awygle** [comment]
I've made the changes we discussed for Windows support. @cr1901 , can you test this to make sure it works for your usecase? It did the right thing as I understand it for both of my msys Pythons, as well as my two Cygwin Pythons, but you probably have a more complete understanding of the issue.

**Comment by olofk** [comment]
Looks great. I'll push it as soon as I hear the results from @cr1901's smoke test

**Comment by cr1901** [comment]
@olofk @awygle Worked fine. No issues downloading from OpenCores with @awygle's fix, which tbh I prefer to mine.

```
DEBUG: Using cygpath translation
DEBUG: None
DEBUG:     svn co -q --no-auth-cache -r 3 --username orpsoc --password orpsoc http://opencores.org/ocsvn/jt51/jt51/trunk /home/William/.cache/fusesoc/jt51_0
```

Second line with `None` reflects the `cwd` argument to the launcher. I assume that's okay?

**Comment by olofk** [comment]
Awesome. Thanks for patches and testing. Picked and pushed!
